app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java

```diff
@@ -9,23 +10,156 @@
+                    "  \n" +
+                    "  # Get the label in the preferred language of the user, or any other language if no label is available in that language.\n" +
+                    "  OPTIONAL {?item rdfs:label ?item_label_preferred_language. FILTER (lang(?item_label_preferred_language) = \"{{LANG}}\")}\n" +
+                    "  OPTIONAL {?item rdfs:label ?item_label_any_language}\n" +
+                    "  \n" +
+                    "  # Get the class label in the preferred language of the user, or any other language if no label is available in that language.\n" +
+                    "  OPTIONAL {\n" +
+                    "    ?item p:P31/ps:P31 ?classId.\n" +
+                    "    OPTIONAL {?classId rdfs:label ?class_label_preferred_language. FILTER (lang(?class_label_preferred_language) = \"{{LANG}}\")}\n" +
+                    "    OPTIONAL {?classId rdfs:label ?class_label_any_language}\n" +
+                    "\n" +
+                    "    # Get icon\n" +
+                    "    OPTIONAL { ?classId wdt:P2910 ?icon0. }\n" +
+                    "    OPTIONAL { ?classId wdt:P279*/wdt:P2910 ?icon1. }\n" +
+                    "    # Get emoji\n" +
+                    "    OPTIONAL { ?classId wdt:P487 ?emoji0. }\n" +
+                    "    OPTIONAL { ?classId wdt:P279*/wdt:P487 ?emoji1. }\n" +
+                    "  }\n" +
+                    "}\n" +
+                    "GROUP BY ?item\n";
+    private static List<Place> places;
+    private static NearbyPlaces singleton;
+    private double radius = RADIUS_INITIAL;
+
+    private NearbyPlaces(){
+    }
 
-    public static List<Place> get() {
-        if(places != null) {
-            return places;
+    public List<Place> getFromWikidataQuery(LatLng curLatLng, String lang) {
+        List<Place> places = Collections.emptyList();
+        while (true) {
+            places = getFromWikidataQuery(curLatLng, lang, radius);
+            if (places.size() >= MIN_RESULTS) {
+                break;
+            } else {
+                radius *= RADIUS_MULTIPLIER;
+            }
+        }
+        while (places.size() > MAX_RESULTS && radius > RADIUS_INITIAL) {
+            radius /= RADIUS_MULTIPLIER;
+            places = getFromWikidataQuery(curLatLng, lang, radius);
+        }
+        Log.d(TAG, "current radius: " + radius);
+        return places;
+    }
+
+    private static List<Place> getFromWikidataQuery(LatLng curLatLng, String lang, double radius) {
+        try {
+            String query = WIKIDATA_QUERY_TEMPLATE.replace("{{RADIUS}}", "" + radius)
+                    .replace("{{LAT}}", "" + curLatLng.latitude)
+                    .replace("{{LONG}}", "" + curLatLng.longitude)
+                    .replace("{{LANG}}", "" + lang);
+            query = URLEncoder.encode(query, "utf-8");
+            String url = WIKIDATA_QUERY_URL.replace("{{QUERY}}", query);
+            Log.v(TAG, url.toString());
```

Comment:
Would you mind logging the radius too? I guess fine-tuning the radius one of the most important reasons to read the log, and right now it is very hard to read from that long query string. Thanks a lot!